### PR TITLE
DM-45899: Write a Task to compute Ex for TEx

### PR DESCRIPTION
**Pull request by PFLeget**
See https://rubinobs.atlassian.net/browse/DM-45899 for more informations.

### CODE REVIEW

File: python/lsst/meas/algorithms/__init__.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/__init__.py:65]: Let's avoid this. Older modules are imported by default to not break things, but newer modules are expected to be specifically imported if required.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:34]: We should put this as its own file. We anticipate reusing this in `analysis_tools` and it is not evident which module to find this in if we put it here.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:299]: Do we need to do a deepcopy here? Does `kk_E1` actually get overridden when we process the next catalog?

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:305]: There is some awkwardness in starting a variable with upper case (as opposed to a class). I don't have a better suggestion at the moment.

File: tests/test_computeExPsf.py
Comment (arunkannawadi) [inline tests/test_computeExPsf.py:259]: Could we include some intermediate scale as well? It'd be at least useful to check that the values from `E2` is about twice more than `E1` (or whatever factor we naively expect).

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:280]: I suspect we'd want to just take a catalog and extract these columns here. But that can wait until we start using this.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:39]: It'd be good to check if the defaults are consistent with the  version of `treecorr` available in the `rubin-env` / shared stack at USDF.

**Review by arunkannawadi — COMMENTED** [review]
There's actually a guide on how to transfer code between packages: https://developer.lsst.io/stack/transferring-code.html#transferring-code-between-packages

Although the history is minimal, it'd be a good practice to follow this procedure.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (PFLeget) [reply python/lsst/meas/algorithms/computeExPsf.py:299]: Yes and yes. I was surprised by that but this is what happened while I was doing unit tests. If you have another suggestions, please let me know.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (PFLeget) [reply python/lsst/meas/algorithms/computeExPsf.py:305]: I guess this is just because in all tickets TEx is upper case and I guess putting lower case can introduce confusion with ellipticity  parameters. Open to suggestions and don't have a big feeling on this.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (PFLeget) [reply python/lsst/meas/algorithms/computeExPsf.py:39]: I did and I made the corrections if needed.

File: python/lsst/meas/algorithms/treecorrUtils.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/treecorrUtils.py:1]: The standard preamble is missing.

File: python/lsst/meas/algorithms/treecorrUtils.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/treecorrUtils.py:19]: It's time to generalize this and replace 'CalcRhoStatistics class' with correlations of PSF residuals or something similar.

File: python/lsst/meas/algorithms/treecorrUtils.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/treecorrUtils.py:20]:
```suggestion
    some of the fields that are relevant for rho-statistics calculations and the likes of it.
```

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:31]: THis should be a list or tuple.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:70]: This `Task` needs to have a corresponding `ConfligClass`, that _has_ a `ConfigurableField` of type `TreecorrConfig`.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [reply python/lsst/meas/algorithms/computeExPsf.py:103]: One option would be to call them `metric_E1` instead of `E1`.

**Review by arunkannawadi — COMMENTED** [review]
The transfer branch should move the code as it was. Suggested doc changes and updating the default values to correspond to treecorr `v4.3` should happen after the transfer branch is merged to the ticket branch.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (PFLeget) [reply python/lsst/meas/algorithms/computeExPsf.py]: I guess I did what you described, but not 100% sure. If you can have a look at what I did on that, I won't mind.

Comment (arunkannawadi) [reply python/lsst/meas/algorithms/computeExPsf.py]: You'll need to follow the pattern similar to the `Config` for `CalcRhoStatistics` here: https://github.com/lsst/analysis_tools/blob/dda7fc77d96733815aa073d4c1dc6caacc4ec067/python/lsst/analysis/tools/actions/vector/calcRhoStatistics.py#L161

Comment (arunkannawadi) [reply python/lsst/meas/algorithms/computeExPsf.py]: Ah, I missed the new changed. You might have it all now.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:38]: I understand it can be confusing at first, but make this just `treecorr`.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:96]: We should check that `units` here match up with the `units` in `TreecorrConfig`.

**Review by arunkannawadi — COMMENTED** [review]
Are the default values in `TreecorrConfig` suitable for the metric calculation? I don't quite remember what binning values we settled on, and what the default ones would be.  `ComputeExPsfConfig` should probably have a `setDefaults` method that would set it to the `nbins`, `min_sep`, `max_sep` and `units` values appropriately.

File: python/lsst/meas/algorithms/computeExPsf.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/computeExPsf.py:42]: I suspect this needs an explicit `validate` method that calls `self.treecorr.validate`. Because `TreecorrConfig` is not a `ConfigClass` of a sub task, I am fairly certain that it will never be validated.

Comment (arunkannawadi) [reply python/lsst/meas/algorithms/computeExPsf.py:42]: There also needs to be a `setDefaults` method here that sets the `min_sep`, `max_sep` and either `nbins` or `bin_size` as appropriate here corresponding to the requirements. @jmeyers314 - what was the binning we settled on finally for TEx computations?

Comment (PFLeget) [reply python/lsst/meas/algorithms/computeExPsf.py:42]: For the `validate`, I tried in the unit test and putting wrong value in `ComputeExPsfConfig` (`min_sep>max_sep`) for example and it triggered the `validate` from `TreecorrConfig`. Do I need to do more than that?

Comment (arunkannawadi) [reply python/lsst/meas/algorithms/computeExPsf.py:42]: I think it'd be best to put it here, given that i)  we might want to run this dynamically outside of `computeSummaryStats` and ii) it makes sense for this class to have its own defaults for `treecorr` rather than `computeSummaryStats` setting them.

**Review by arunkannawadi — APPROVED** [review]
Same comment as in `analysis_tools` PR - start the commit messages with upper case. Good to go after you set the defaults here in `setDefaults`.

File: python/lsst/meas/algorithms/treecorrUtils.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/treecorrUtils.py:206]: Post merge comment: This should remain as `rng_seed`. `rng` typically refers to a `np.RandomState` instance, and that's certainly what `treecorr` means. `rng` itself is not nicely persistable and settable as a config file. `rng_seed` makes it explicit that it is the integer value used to seed that instance.

File: python/lsst/meas/algorithms/treecorrUtils.py
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/treecorrUtils.py:209]: Post merge comment: Restore the previous default value and remove `optional=True`.

**Pull request merged**
